### ISSUE TITLE
Memory leak with many small vectors of vectors

### ISSUE THREAD

**Issue by Byrth** [JuliaLang/julia#42566]
This problem was [originally reported in the helpdesk Slack](https://julialang.slack.com/archives/C6A044SQH/p1633704159387700) against DataFrames, but I believe I have replicated it just using vectors of vectors.

OP was using Julia 1.5.0 with DataFrame 1.2.2 on Linux
I used Julia 1.6.2 on WSL2 Ubuntu

Just using vectors, everything behaves as expected and julia's allocated memory does not increase:
```julia
function inner_df(Nrow,Ncol) # Create small dataframe
    rand(Nrow*Ncol)
end

function outer_df(N) #Stack small dataframes
    Nrow=76
    Ncol=21
    df = Vector{Float64}(undef,0)
    for i = 1:N
        append!(df,inner_df(Nrow,Ncol))
    end
    return df
end

function iterated(Niter) # Create a large DataFrame many times.
    for i =1:Niter
        println(i)
        @time pan = outer_df(1000*10*10)
    end
    nothing
end


# (code and top lines after the code)
iterated(1)
GC.gc()
# 12142 user      20   0 1892304 215296  64744 S   0.0   1.3   0:19.41 julia
iterated(10)
GC.gc()
# 12142 user      20   0 1892416 226740  65412 S   0.0   1.4   1:51.96 julia
iterated(10)
GC.gc()
# 12142 user      20   0 1892416 227000  65416 S   0.0   1.4   3:25.04 julia
iterated(20)
GC.gc()
# 12142 user      20   0 1892416 227104  65416 S   0.0   1.4   6:32.83 julia
```

Using vectors of vectors, there seems to be a memory leak:
```julia
function inner_df(Nrow,Ncol) # Create small dataframe
    [rand(Nrow) for i in 1:Ncol]
end

function outer_df(N) #Stack small dataframes
    Nrow=76
    Ncol=21
    df=[Vector{Float64}(undef, 0) for i in 1:Ncol]
    for i = 1:N
        df2 = inner_df(Nrow,Ncol)
        for j in 1:Ncol
            append!(df[j],df2[j])
        end
    end
    return df
end

function iterated(Niter) # Create a large DataFrame many times.
    for i =1:Niter
        println(i)
        @time pan = outer_df(1000*10*10)
    end
    nothing
end


# (code and top lines after code)
iterated(1)
GC.gc()
#  9204 user      20   0 2275056 216848  64844 S   0.0   1.3   0:15.61 julia
iterated(10)
GC.gc()
#  9204 user      20   0 3339728 952280  64844 S   0.0   5.7   1:06.03 julia
iterated(10)
GC.gc()
#  9204 user      20   0 3339728 952876  64844 S   0.0   5.7   1:55.75 julia
iterated(20)
GC.gc()
#  9204 user      20   0 3343824 953112  64964 S   0.0   5.7   3:31.63 julia
```

Please feel free to close if this is a vagary of Linux memory management.

**Comment by LilithHafner** [comment]
I'm on  a mac (11.4) and for this comment I'm using the standard Julia REPL from 1.6.1 in a standard mac terminal.

I'm not sure what `#  9204 user      20   0 3343824 953112  64964 S   0.0   5.7   3:31.63 julia` means, but tried to replicate it on my machine.

```julia
#fresh session
#95 MB
#your function declarations
#116 MB
iterated(1)
#1.81 GB
GC.gc()
#631 MB
iterated(10)
#1.47 GB
GC.gc()
#648 MB
```
And in another way
```julia
#Fresh julia session
#95 MB
function inner_df(Nrow,Ncol) # Create small dataframe
           [rand(Nrow) for i in 1:Ncol]
       end
function outer_df(N, Nrow, Ncol) #Stack small dataframes
           df=[Vector{Float64}(undef, 0) for i in 1:Ncol]
           for i = 1:N
               df2 = inner_df(Nrow,Ncol)
               for j in 1:Ncol
                   append!(df[j],df2[j])
               end
           end
           return df
       end
function iterated(Niter, N, Nrow, Ncol) # Create a large DataFrame many times.
           for i =1:Niter
               println(i)
               @time pan = outer_df(N, Nrow, Ncol)
           end
           nothing
       end
#128 MB
iterated(1,1,1,1); GC.gc()
#134 MB
iterated(10,100_000,76,11); GC.gc()
#621 MB
iterated(10,100_000,76,11); GC.gc()
#621 MB
```
And in another way
```julia
#Fresh julia session
#95 MB
function inner_df(Nrow,Ncol) # Create small dataframe
           [rand(Nrow) for i in 1:Ncol]
       end
function outer_df(N, Nrow, Ncol) #Stack small dataframes
           df=[Vector{Float64}(undef, 0) for i in 1:Ncol]
           for i = 1:N
               df2 = inner_df(Nrow,Ncol)
               for j in 1:Ncol
                   append!(df[j],df2[j])
               end
           end
           return df
       end
#125 MB
outer_df(1_000, 67, 21); GC.gc()
#149 MB
outer_df(10_000, 67, 21); GC.gc()
#160 MB
outer_df(100_000, 67, 21); GC.gc()
#465 MB
outer_df(300_000, 67, 21); GC.gc()
#821 MB
outer_df(300_000, 67, 21); GC.gc()
#822 MB
1+1; GC.gc()
#822 MB
```
I conclude that
-While I am not expert in memory management, there seems to be a memory leak
-It's size depends on N
-Iteration does not result in increased memory use

I hope to look into this and try to simplify the MWE more later today.

**Comment by eirikbrandsaas** [comment]
Well, at least it's great to see that I'm not the only one with this issue. Cross-referencing to the issue I created on DataFrames.jl  (https://github.com/JuliaData/DataFrames.jl/issues/2902).

On my end I have checked using Julia 1.5.0, 1.5.3, 1.6.1 and using two different computers (one with Ubuntu 18.12 and one with a 20.04 Ubuntu-clone (Pop!_os).

**Comment by LilithHafner** [comment]
What I have been able to reproduce is not a leak _per se_, because repeated actions do not increase memory usage, but it is nonetheless large memory usage where I think there should be much less. 

Has anyone been able to reproduce a leak which involves a linear growth in unreleased memory with respect to iterations of a supposedly pure function?

**Comment by vchuravy** [comment]
What happens if you use `GC.gc(true); GC.gc()`.

**Comment by LilithHafner** [comment]
Yes! It does seem like a memory release to system problem.

> GC.gc(true); GC.gc()
Does not fix it.

@JeffBezanson is it possible to call `malloc_trim` manually without rebuilding Julia? I tried `ccall(:malloc_trim, Int32, (Int32,), 0)` and `ccall((:malloc_trim, "malloc"), Int32, (Int32,), 0)`, but neither worked, and I imagine Revise.jl doesn't work for editing .c files in Julia's source.

**Comment by IanButterworth** [comment]
I've been seeing a long running CI run getting OOM killed and it appears to be related to this.

All the tests are enclosed in `@testset`s and julia appears to think the memory is being freed.

Periodically monitoring
```
gc_live = Base.format_bytes(Base.gc_live_bytes())
pid_mem = Base.format_bytes(parse(Int, strip(read(`ps -o rss= $(getpid())`, String))) * 1024)
```
I see the following profile over time (this is just before the run was killed)

<img width="482" alt="Screen Shot 2022-02-02 at 9 43 28 PM" src="https://user-images.githubusercontent.com/1694067/152279858-34c21b68-0fc3-4403-ad04-dfd2ecaf25a2.png">

I tried the [heap snapshot PR](https://github.com/JuliaLang/julia/pull/42286) rebased on master, but it's currently segfaulting

**Comment by IanButterworth** [comment]
Ok, I got the heap snapshot to work
```
julia> open("before.heapsnapshot", "w") do io
           GC.take_heap_snapshot(io)
       end

julia> [zeros(10) for _ in 1:10^7];

julia> GC.gc(true)

julia> open("after.heapsnapshot", "w") do io
           GC.take_heap_snapshot(io)
       end
```

| Before (julia 162 MB)      | After (julia 1.52 GB) |
| ----------- | ----------- |
|  <img width="699" alt="Screen Shot 2022-02-02 at 11 51 33 PM" src="https://user-images.githubusercontent.com/1694067/152283412-f6b64bbc-9250-42f1-ba2d-710a91d35c56.png"> | <img width="696" alt="Screen Shot 2022-02-02 at 11 51 19 PM" src="https://user-images.githubusercontent.com/1694067/152283429-17094ed4-cb7c-4946-bb2e-bb260b457bb7.png"> |


I don't see the memory accounted for there

**Comment by racinmat** [comment]
> Yes! It does seem like a memory release to system problem.
> 
> > GC.gc(true); GC.gc()
> 
> Does not fix it.
> 
> @JeffBezanson is it possible to call `malloc_trim` manually without rebuilding Julia? I tried `ccall(:malloc_trim, Int32, (Int32,), 0)` and `ccall((:malloc_trim, "malloc"), Int32, (Int32,), 0)`, but neither worked, and I imagine Revise.jl doesn't work for editing .c files in Julia's source.

In the https://github.com/JuliaLang/julia/issues/30653#issuecomment-481940653 they call it as `ccall(:malloc_trim, Cvoid, (Cint,), 0)` would that work for you, or still nothing?

**Comment by IanButterworth** [comment]
I modified the source to call that for every GC.gc already. 

But.. it was just pointed out to me that MacOS doesn't use glibc so perhaps I should have done something similar but for MacOS.

**Comment by gbaraldi** [comment]
There is no malloc_trim in macos libc, at least man and google tell me so, apparently it's a GNU specific extension.

**Comment by IanButterworth** [comment]
Another datapoint. This is with MacOS's instruments running 
`julia -e "x = [zeros(10) for _ in 1:10^7]; sleep(5); x = nothing; GC.gc(true); sleep(5);"`

It appears to only know about a few hundred MiB, but the process rises to 1.5 GB and stays there

<img width="1111" alt="Screen Shot 2022-02-03 at 9 56 26 AM" src="https://user-images.githubusercontent.com/1694067/152367581-3f2a53b4-1183-40ea-93a9-7d283ab81507.png">

**Comment by paulmelis** [comment]
On Arch Linux, with 1.7.1 I get (based on your earlier examples):

```
melis@juggle 16:02:~/$ cat array.jl 
# https://github.com/JuliaLang/julia/issues/42566#issuecomment-1028603528

function info(s)
    gc_live = Base.format_bytes(Base.gc_live_bytes())
    pid_mem = Base.format_bytes(parse(Int, strip(read(`ps -o rss= $(getpid())`, String))) * 1024)    
    println("$(s) | gc_live = $(gc_live), pid_mem = $(pid_mem)"); 
    readline()    
end

info("about to array alloc")
x = [zeros(10) for _ in 1:10^7];

info("about to run GC")
x = nothing
GC.gc(true)

info("about to exit")
melis@juggle 16:02:~/c/julia-git$ julia array.jl
about to array alloc | gc_live = 28.629 MiB, pid_mem = 199.867 MiB

about to run GC | gc_live = 1.417 GiB, pid_mem = 1.627 GiB

about to exit | gc_live = 1.749 MiB, pid_mem = 207.352 MiB
```

So could be a MacOS specific thing, as I don't see the large process size at exit on Linux?

**Comment by IanButterworth** [comment]
https://github.com/JuliaLang/julia/issues/42566#issuecomment-1028587020 was on ubuntu (just clarified in the comment)

**Comment by paulmelis** [comment]
> [#42566 (comment)](https://github.com/JuliaLang/julia/issues/42566#issuecomment-1028587020) was on ubuntu (just clarified in the comment)

Okay, but I guess that was with much more extensive code than the simple array allocation above? And possibly multi-threaded? Seems good to reduce simple test cases, hence what I observed on Linux compared to your MacOS.

**Comment by gbaraldi** [comment]
I was doing some reading and apparently macos is very lazy on returning memory to the kernel, if the memory pressure increases it does do it, but only then. 

On an ARM ubuntu server the original OP sometimes shows that the OS hasn't requested the memory back, while other times it does get it back. I imagine it's different behaviour from the libcs

**Comment by IanButterworth** [comment]
Ok. Adding `ccall(:malloc_trim, Cvoid, (Cint,), 0)` to my Ubuntu CI run is not freeing up the memory. So something else is going on there. The code _is_ heavily multithreaded, also uses CUDA, so clearly needs reduction.

**Comment by racinmat** [comment]
Just cross-referencing, in my hunt for freeing more memory I learned the `malloc_trim` and it helped partly, it cleared some memory which has not been cleared before, but not everything. https://discourse.julialang.org/t/poor-performance-of-garbage-collection-in-multi-threaded-application/75493/18

**Comment by dlakelan** [comment]
This issue might be related to the behavior I reported on the discourse here:

https://discourse.julialang.org/t/determining-size-of-dataframe-for-memory-management/85277/36

MWE is from reading in and sub-sampling some big CSV files available from the Census:

```
using Pkg
Pkg.activate(".")

using CSV,DataFrames,Downloads,DataFramesMeta, StatsPlots, XLSX, 
    Dates, Statistics, Turing, LinearAlgebra, Interpolations, Serialization,
    GLM, Colors, ColorSchemes



function getifnotthere(filename,URL)
    if !Base.Filesystem.ispath(filename)
        Downloads.download(URL,filename)
    end
end


getifnotthere("data/pums-2020-5yr-hus.zip","https://www2.census.gov/programs-surveys/acs/data/pums/2020/5-Year/csv_hus.zip")


if ! Base.Filesystem.ispath("data/psam_husa.csv") || ! Base.Filesystem.ispath("data/psam_husb.csv")
    cd("data")
    run(`unzip pums-2020-5yr-hus.zip`)
    cd("..")
end

psamh = let psamh = DataFrame()
    for i in ["a","b","c","d"]
        new = @select(CSV.read("data/psam_hus$(i).csv",DataFrame),:SERIALNO,:ST,:FINCP,:NP,:WGTP)
        @subset!(new,.!ismissing.(:FINCP) .&& .! ismissing.(:NP))
        psamh = [psamh ; 
            new[wsample(1:nrow(new),new.WGTP,round(Int64,0.1*nrow(new))),:] ]
    end
    psamh
end
psamh.Year = tryparse.(Int64,psamh.SERIALNO[i][1:4] for i in 1:nrow(psamh))

```
At the end of this I have like 10GB of memory usage by the Julia process but varinfo() only knows about ~100MB of variables. Eventually systemd-oomd kills my session.

**Comment by baggepinnen** [comment]
>This issue might be related to the behavior I reported on the discourse here:
https://discourse.julialang.org/t/determining-size-of-dataframe-for-memory-management/85277/36
MWE is from reading in and sub-sampling some big CSV files available from the Census: 

See
https://github.com/JuliaData/CSV.jl/issues/850

**Comment by dpinol** [comment]
Upstream bug has been closed https://sourceware.org/bugzilla/show_bug.cgi?id=27103#c2

> This particular issue is not a bug, the allocator is under no requirement to free back the 50MiB to the OS if it believes that the memory will be used again and that performance is improved by caching it in userspace.

On my ubuntu, neither `ccall(:malloc_trim, Cvoid, (Cint,), 0)` nor `ccall(:malloc_trim, Int32, (Int32,), 0)` reclaimed the memory used by `[zeros(10) for _ in 1:10^7]`

**Comment by elextr** [comment]
I havn't been following the discussion in detail, but have encountered it elsewhere.

It seems to me that some correspondents seem to miss the point that _any_ allocation located above blocks of free memory will prevent `malloc_trim()` from returning the free blocks, even if all the large block(s) have been `free()`d.  So an allocation by the repl, the compiler, gc itself, anything, can block all or some memory from getting returned by `malloc_trim()`.

Also in the case of multiple threaded allocations `malloc_trim()` only trims the main arena, not any additional arenas created by glibc to minimise lock contention with the multiple threads allocating.

tl;dr `malloc_trim()` is not a panacea.

**Comment by dlakelan** [comment]
> This particular issue is not a bug, the allocator is under no requirement to free back the 50MiB to the OS if it believes that the memory will be used again and that performance is improved by caching it in userspace.

This attitude is really flawed. In my example the systemd-oomd is killing Julia because it's taking 75% of the system RAM even though only 1% of that ram is actually in use by Julia.

**Comment by BioTurboNick** [comment]
Yeah, this is a problem. At a minimum, it should be possible to deliberately ask Julia to compact its memory. I'm currently looking at an 8 GB process. All it does it load data and save it; it shouldn't retain any references. `ccall(:malloc_trim, Int32, (Int32,), 0)` reduced it to 1.7 GB.

The memory might be needed again in a day or a week, but not immediately.

**Comment by BioTurboNick** [comment]
> @BioTurboNick Could you try #47062 in your workload? And maybe compare the performance(runtime and max memory)?

@gbaraldi what would I need to do?

**Comment by gbaraldi** [comment]
Just build julia on my PR and try your workload ;)

**Comment by juliohm** [comment]
I just reported another instance of a similar memory leak with NearestNeighbors.jl:

https://github.com/KristofferC/NearestNeighbors.jl/issues/208

The attached MWE leaks 30Gb of memory. Posterior calls to `Base.GC.gc()` don't release memory.

Like in this issue here, the "vector of vectors" type plays a role. The leak only occurs on Linux.

**Comment by oscardssmith** [comment]
I think this issue may be slightly unrelated as testing with
```
using NearestNeighbors, StaticArrays
x(i, N) = float.(Tuple(CartesianIndices((N, N, N))[i]))

function mwe(N)
    sources = rand(1:N^3, 3000) |> unique
    others = setdiff(1:N^3, sources)
    batches = Iterators.partition(others, 10^3)
    kdtree = KDTree([SVector(x(i, N)...) for i in sources])
    dists = []
    for batch in batches
        xs = [SVector(x(i, N)...) for i in batch]
        _, ds = knn(kdtree, xs, length(sources), true)
        append!(dists, ds)
    end
end
```

shows that calling this once leaks a couple hundred mb, but calling it repeatedly doesn't leak more.

```
julia> @time mwe(45)
 24.323221 seconds (529.90 k allocations: 3.910 GiB, 2.06% gc time)

julia> gc_live = Base.format_bytes(Base.gc_live_bytes())
"2.083 GiB"

julia> pid_mem = Base.format_bytes(parse(Int, strip(read(`ps -o rss= $(getpid())`, String))) * 1024)
"2.855 GiB"

julia> @time mwe(45)
 26.283792 seconds (529.91 k allocations: 3.910 GiB, 2.33% gc time)

julia> @time mwe(45)
 27.979254 seconds (529.87 k allocations: 3.920 GiB, 2.01% gc time)

julia> gc_live = Base.format_bytes(Base.gc_live_bytes())
"2.110 GiB"

julia> pid_mem = Base.format_bytes(parse(Int, strip(read(`ps -o rss= $(getpid())`, String))) * 1024)
"3.025 GiB"

julia> @time mwe(45)
 24.051036 seconds (529.93 k allocations: 3.899 GiB, 2.70% gc time)


julia> gc_live = Base.format_bytes(Base.gc_live_bytes())
"2.078 GiB"

julia> pid_mem = Base.format_bytes(parse(Int, strip(read(`ps -o rss= $(getpid())`, String))) * 1024)
"3.025 GiB"

julia> ccall(:malloc_trim, Cvoid, (Cint,), 0)

julia> pid_mem = Base.format_bytes(parse(Int, strip(read(`ps -o rss= $(getpid())`, String))) * 1024)
"2.803 GiB"
```

so it looks like something about this MWE is allocating global state somewhere (possibly as a cache?)

**Comment by juliohm** [comment]
As we discussed on Zulip, this only happens on Linux. The function `mwe` as originally shared in that NearestNeighbors.jl issue allocates up to 30Gb and never frees it. I am calling `Base.GC.gc()` after the `mwe()` call and the 30Gb are not freed either.

Your MWE above is slightly different than the one I shared. In particular you are calling the function with `N=45`, but my original MWE always uses `N=100*100*100`.
